### PR TITLE
[SPARK-50157][SQL] Using SQLConf provided by SparkSession first.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
This PR proposes to use `SQLConf` provided by `SparkSession` first.

### Why are the changes needed?
`SQLConf` provided by `SparkSession` have better perf than `SQLConf.get`.

### Does this PR introduce _any_ user-facing change?
'No'.

### How was this patch tested?
GA tests.
The benchmark test.
```
object SQLConfBenchmark extends SqlBasedBenchmark {
  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    runBenchmark("Get SQLConf") {
      val iters = 1000
      val benchmark = new Benchmark("Benchmark SQLConf", iters, output = output)
      benchmark.addCase("SQLConf.get") { _ =>
        for (_ <- 1 to iters) {
          val conf = SQLConf.get
        }
      }

      benchmark.addCase("sessionState.conf") { _ =>
        for (_ <- 1 to iters) {
          val conf = spark.sessionState.conf
        }
      }

      benchmark.run()
    }
  }
}
```
The benchmark output.
```
Benchmark SQLConf:                        Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
SQLConf.get                                           0              0           1         22.9          43.6       1.0X
sessionState.conf                                     0              0           0       1377.4           0.7      60.1X
```

### Was this patch authored or co-authored using generative AI tooling?
'No'.

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
> It would be great if you can provide some concrete examples or numbers with this, @beliefer .

In fact, `SQLConf.get` and `spark.sessionState.conf` are functionally equivalent. But `spark.sessionState.conf` is faster than `SQLConf.get` because the latter requires visit thread local. So we can see the usage scenarios:

1. The code rely on `SQLConfHelper` in Spark we can't find `SparkSession` in context.
2. All the places in Spark rely on `spark.sessionState.conf` if exists `SparkSession`.

It was hard to provide some perf number for me, the perf improvement looks very inconspicuous after all.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala [hunk 1]
```diff
@@ -231,7 +231,7 @@ class HiveQuerySuite extends HiveComparisonTest with SQLTestUtils with BeforeAnd
   createQueryTest("no from clause",
     "SELECT 1, +1, -1")
 
-  if (!conf.ansiEnabled) {
+  if (!spark.sessionState.conf.ansiEnabled) {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveQuerySuite.scala:234]: Kindly revert all modifications made to this test suite, as the test code is inconsequential to the performance evaluation.

**Comment by beliefer** [comment]
@dongjoon-hyun Please take a look again.

**Comment by HyukjinKwon** [comment]
How much does it get improved?

**Comment by beliefer** [comment]
> How much does it get improved?

The benchmark added into PR's description.

**Review by dongjoon-hyun — COMMENTED** [review]
This PR technically proposed to avoid the use of `SQLConfHelper.conf` method, right?

If that method is useless, can we remove the method? Or, when do we need to use that method still?

**Comment by HyukjinKwon** [comment]
Does the benchmark result say `session.conf` is slower than `SQLConf.get`?

**Comment by beliefer** [comment]
> Does the benchmark result say `session.conf` is slower than `SQLConf.get`?

`session.conf` is quicker than `SQLConf.get`.

**Comment by cloud-fan** [comment]
In general I think it's better to use the `SQLConf` in `SparkSession` if it's available. Maybe we should just override `def conf` in those rules that have `SparkSession` avaliable

**Comment by beliefer** [comment]
@cloud-fan `override def conf` finished, do you have any other comments?

**Comment by beliefer** [comment]
@dongjoon-hyun @HyukjinKwon The GA failure seems unrelated. Please take a look again.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala [hunk 2]
```diff
@@ -33,6 +33,8 @@ import org.apache.spark.util.Utils
  */
 case class CoalesceShufflePartitions(session: SparkSession) extends AQEShuffleReadRule {
 
+  override def conf: SQLConf = session.sessionState.conf
```
Comment (LuciferYang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala:36]: Can we change it to `override val conf: SQLConf = session.sessionState.conf` in cases where `conf` needs to be called multiple times?

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala (hunk 2 above)
Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala]: From the test results, there are other issues with changing to `val`. It seems we need to revert the last commit.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala]: Yes. I think so.

Comment (LuciferYang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CoalesceShufflePartitions.scala]: sorry for misleading

**Review by LuciferYang — APPROVED** [review]

**Comment by beliefer** [comment]
@LuciferYang @dongjoon-hyun @cloud-fan @HyukjinKwon Thank you all.

**Pull request closed without merging**